### PR TITLE
Added list clearing codegen to AOTAutograd (hidden behind config.aot_clear_list

### PR DESCRIPTION
**Pull request by Chillee**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #83248
* #83240
* #83235
* #83215
* __->__ #83137

### CODE REVIEW

**Comment by Chillee** [comment]
cc: @albanD on changes to `python_function.cpp`

cc: @jansel and @SherlockNoMad on changes we need to make to the compilers being passed to AOTAutograd.

**Comment by Chillee** [comment]
image

An example of the changes.

File: torch/csrc/autograd/python_function.cpp [hunk 1]
```diff
@@ -806,6 +806,12 @@ PyObject* THPFunction_name(PyObject* self, PyObject* noargs) {
   END_HANDLE_TH_ERRORS
 }
 
+PyObject* THPFunction_clear_saved(PyObject* self, PyObject* noargs) {
+  HANDLE_TH_ERRORS((THPFunction*)self)->saved_variables.clear();
```
Comment (albanD) [inline torch/csrc/autograd/python_function.cpp:810]: You can move the code to the next line?

File: torch/csrc/autograd/python_function.cpp [hunk 2]
```diff
@@ -1184,6 +1190,7 @@ static struct PyGetSetDef THPFunction_properties[] = {
 // NOLINTNEXTLINE(modernize-avoid-c-arrays,cppcoreguidelines-avoid-c-arrays,cppcoreguidelines-avoid-non-const-global-variables)
 static struct PyMethodDef THPFunction_methods[] = {
     {(char*)"name", THPFunction_name, METH_NOARGS, nullptr},
+    {(char*)"clear_saved", THPFunction_clear_saved, METH_NOARGS, nullptr},
```
Comment (albanD) [inline torch/csrc/autograd/python_function.cpp:1193]: nit: `clear_saved_tensors` to match the field name

File: torch/csrc/autograd/python_function.cpp (hunk 1 above)
Comment (albanD) [inline torch/csrc/autograd/python_function.cpp:810]: Also use `((THPFunction*)self)->cdata->release_variables()` which does the right logic at the Node level.

File: functorch/functorch/_src/config.py [hunk 3]
```diff
@@ -20,6 +20,9 @@
 #   fix for complex numbers
 use_fake_tensor = False
 
+# Changes AOTAutograd to passing a list of tensors that are then cleared
```
Comment (albanD) [inline functorch/functorch/_src/config.py:23]: The user is passing the list?

File: functorch/functorch/_src/aot_autograd.py [hunk 4]
```diff
@@ -238,12 +263,25 @@ def fake_fn(primals, tangents):
                 with track_graph_compiling("joint"):
                     fw_module, bw_module = partition_fn(fx_g, joint_inputs)
 
+                # Used to prove out memory-savings
```
Comment (albanD) [inline functorch/functorch/_src/aot_autograd.py:266]: Should this be left here?

**Review by albanD — COMMENTED** [review]
Small nits only

File: functorch/functorch/_src/config.py (hunk 3 above)
Comment (Chillee) [reply functorch/functorch/_src/config.py:23]: So today, the contract with the compiler is that we pass in a function like
```
f(a, b, c)
```

and the compiler returns a callable
```
compiled_f(a,b,c)
```

This changes the contract so that was pass in a function
```
f([a,b,c])
```
and the compiler returns a callable

```
compiled_f([a,b,c])
```

**Review by jansel — APPROVED** [review]

File: functorch/functorch/_src/config.py (hunk 3 above)
Comment (albanD) [reply functorch/functorch/_src/config.py]: I'm not sure to see the difference? Don't you use pytree to "unpack" any data structure in the args?

Comment (Chillee) [reply functorch/functorch/_src/config.py]: @albanD I'll add some more context to this issue/commit, but basically, it's about object lifetimes.

If you call `f(a,b,c)`, then a, b, and c will stay alive for the duration of `f`, and there's no way around this AFAIK.

OTOH, if you call `f([a,b,c])`, then we can clear the list and free a/b/c inside of `f`.

Comment (albanD) [reply functorch/functorch/_src/config.py]: You can do `del a, b, c` ?

Comment (Chillee) [reply functorch/functorch/_src/config.py]: Unfortunately not :( Gimme a sec.

**Comment by Chillee** [comment]
## So... what is the problem here?

(cc: @albanD , @eellison)? 

Let's say you have a function like this

```
import torch
def get_mem():
    print(f"{torch.cuda.memory_allocated()/1e9} GB")

def f(x):
    del x
    get_mem()
    return None

f(torch.randn(2**30, device='cuda'))
```

What will this print?

Unfortunately for us, the answer is `4.294967296 GB`. This is despite the fact that within `f`, we no longer have any references to a tensor, and outside `f`, we are creating the tensor to directly pass into `f`.

This is because, in Python, when you call a function, the arguments are always *borrowed references*. Thus, they must be kept alive for the duration of the function call. See [this reference](http://davis.lbl.gov/Manuals/PYTHON-2.5.1/ext/ownershipRules.html).

> When you pass an object reference into another function, in general, the function borrows the reference from you

Thus, there is no way for us to free any of the inputs to the function *for the duration of the function*.

So, how do we solve this? Although the inputs *themselves* must stay alive for the duration of the function, there's no guarantee that the references they hold must be kept alive. So, if instead, we pass a list to the function, then clear that list, we can ensure that the tensor is freed in the function. So...

```
def list_f(x):
    val = x[0]
    x.clear()
    del val
    get_mem()
    return None

list_f([torch.randn(2**30, device='cuda')])
```

## Why is this particularly a problem for AOTAutograd?

Now, why is this issue a problem for AOTAutograd? Well, if you think about the signature of the backwards pass, it looks something like...

```
def backwards(activation_0, activation_1, activation_2, ... activation_million, grad_output)
```

In other words, you're passing in a ton of activations to the backwards pass! And as activations constitute a significant part of the memory in any deep learning model, this can lead to higher memory usage in the backwards pass. Let's look at the graph for AOTAutograd today for a model like `resnet18.

image

You can see that in eager, we reach our peak memory in-between forwards + backwards. Then in the backwards pass, we start to free our activations as they're no longer needed, leading to a reduction in memory usage.

With AOTAutograd, however, our memory doesn't drop! In fact, it steadily rises during the backwards pass, as we additionally allocate gradients in addition to all the activations being saved.

So, that's what this PR does. It fixes this problem for AOTAutograd by changing the convention with which we interface with compilers.

File: torch/csrc/autograd/python_function.cpp (hunk 1 above)
Comment (Chillee) [reply torch/csrc/autograd/python_function.cpp]: I don't actually see how to use this function? It seems like `release_variables` is on PyNode, not `THPFunction*`. How do I get a PyNode from a `THPFunction*`?

**Comment by Chillee** [comment]
@albanD I needed to add some more changes to `python_function.cpp` in order to handle `retains_graph` calls properly.

**Comment by ezyang** [comment]
I wouldn't call this "list clearing", no one is going to know what this means. When @jansel needs one-element mutable boxes, he likes to make a fresh class called SomethingBox, this is my preferred convention too. Or we could pun the name based on C++ std::move and call it Move() or MoveBox() or something.

Also, it's really easy to make backends "compatible" with the box calling convention by decorating them with a function that looks for MoveBoxes, unpacks them and then calls the inner function. So I'd also prefer we not add yet another config flag for this.

**Comment by Chillee** [comment]
I agree the name sucks, but I’m not totally sure how we avoid having a config *somewhere* for this (without breaking any existing backends, although maybe that’s fine?), although I suppose it could be at the backend-specific level.

Unfortunately, I’m also not sure we can/should pass in a custom ArgBox class. One original design constraint here was that we 1. Wanted to be able to use the existing fx graph for it (so need to be able to codegen a prologue that handles it), and 2. I wanted to be able to directly pass it to Torchscript.

Perhaps now, we could plausibly drop 2 as a req?

**Comment by ezyang** [comment]
> I agree the name sucks, but I’m not totally sure how we avoid having a config somewhere for this (without breaking any existing backends, although maybe that’s fine?), although I suppose it could be at the backend-specific level.

Yeah, all I'm saying is that it's easy for a backend to say "I don't support this" and automatically do the dumb thing, and all the backends live here so we should rip the bandaid off and update them all. (And if there are too many backends, well, this is a good reason to delete some of the ones that don't matter.)

> Unfortunately, I’m also not sure we can/should pass in a custom ArgBox class. One original design constraint here was that we 1. Wanted to be able to use the existing fx graph for it (so need to be able to codegen a prologue that handles it), and 2. I wanted to be able to directly pass it to Torchscript.

Only (2) seems to hard require lists? But OK, fine, we can keep the current calling convention as is.

**Comment by Chillee** [comment]
I'm gonna keep the config option for now for the sake of making this easier to land (otherwise I need to make a simultaneous change to torchdynamo repo).

**Comment by ezyang** [comment]
BTW, if you're serious about TS compatibility, wrapping all the args in a list is only ok if we can guarantee that you have a flat argument list of tensors; otherwise you have a heterogenous list which wouldn't be representable in TS type system.

File: torch/csrc/autograd/python_function.cpp [hunk 5]
```diff
@@ -806,6 +806,18 @@ PyObject* THPFunction_name(PyObject* self, PyObject* noargs) {
   END_HANDLE_TH_ERRORS
 }
 
+PyObject* THPFunction_maybe_clear_saved_tensors(
+    PyObject* self,
+    PyObject* noargs) {
+  HANDLE_TH_ERRORS;
+  auto cdata = ((THPFunction*)self)->cdata.lock();
+  if (!cdata->retain_variables) {
```
Comment (albanD) [inline torch/csrc/autograd/python_function.cpp:814]: This is a not a thing I'm afraid.
I think you will have to use the same approach as https://github.com/pytorch/pytorch/pull/82544

**Review by albanD — APPROVED** [review]
Autograd change sounds good!
